### PR TITLE
Cause filter fix

### PR DESCRIPTION
**Pull request by eyalfa**
found a couple of bugs in `Cause.filter(..)` while hacking on something completely different...
fix seemed quite easy, so here goes...

### CODE REVIEW

File: core/shared/src/main/scala/zio/Cause.scala [hunk 1]
```diff
@@ -822,25 +822,42 @@ object Cause extends Serializable {
...
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
       def bothCase(context: Any, left: Cause[E], right: Cause[E]): Cause[E] =
-        if (p(left)) {
-          if (p(right)) Cause.Both(left, right)
+        if (left.nonEmpty) {
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/Cause.scala]: What's the rationale behind this change? Shouldn't we use the predicate when deciding if we're going to include left or right?

File: core/shared/src/main/scala/zio/Cause.scala [hunk 2]
```diff
@@ -822,25 +822,42 @@ object Cause extends Serializable {
...
+          if (right.nonEmpty) Cause.Both(left, right)
           else left
-        } else if (p(right)) right
+        } else if (right.nonEmpty) right
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/Cause.scala]: `nonEmpty` is not necessarily cheap, and it's called at least once, potentially twice per method invocation. Should we store it in vals at the start of the method for both left and right?

File: core/shared/src/main/scala/zio/Cause.scala [hunk 3]
```diff
@@ -822,25 +822,42 @@ object Cause extends Serializable {
     final case class Filter[E](p: Cause[E] => Boolean) extends Folder[Any, E, Cause[E]] {
       def empty(context: Any): Cause[E] = Cause.empty
 
-      def failCase(context: Any, error: E, stackTrace: StackTrace): Cause[E] = Cause.Fail(error, stackTrace)
+      def failCase(context: Any, error: E, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Fail(error, stackTrace)
```
Comment (eyalfa) [inline core/shared/src/main/scala/zio/Cause.scala:827]: @kyri-petrou , this is (one of the places) where the bug is.
the tests proves this, notice the original code did not test the predicate at this point which results with `filter(_ => false)` returning the cause as is.

File: core/shared/src/main/scala/zio/Cause.scala [hunk 4]
```diff
@@ -822,25 +822,42 @@ object Cause extends Serializable {
...
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
-      def dieCase(context: Any, t: Throwable, stackTrace: StackTrace): Cause[E] = Cause.Die(t, stackTrace)
+      def dieCase(context: Any, t: Throwable, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Die(t, stackTrace)
```
Comment (eyalfa) [inline core/shared/src/main/scala/zio/Cause.scala:835]: second incarnation of the same bug

File: core/shared/src/main/scala/zio/Cause.scala [hunk 5]
```diff
@@ -822,25 +822,42 @@ object Cause extends Serializable {
...
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
-      def interruptCase(context: Any, fiberId: FiberId, stackTrace: StackTrace): Cause[E] =
-        Cause.Interrupt(fiberId, stackTrace)
+      def interruptCase(context: Any, fiberId: FiberId, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Interrupt(fiberId, stackTrace)
```
Comment (eyalfa) [inline core/shared/src/main/scala/zio/Cause.scala:843]: third...

File: core/shared/src/main/scala/zio/Cause.scala (hunk 1 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/Cause.scala]: `left` and `right` are already filtered at this point

File: core/shared/src/main/scala/zio/Cause.scala (hunk 2 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/Cause.scala]: agreed, though not sure this is a real concern.
anyways, it can be optimized by using some kind of short circuiting traversal over the `Cause` or alternatively by caching the result of `isEmpty` (which actually do use a short circuiting traversal)

Comment (eyalfa) [reply core/shared/src/main/scala/zio/Cause.scala]: @kyri-petrou , actually in this case we can accumulate the `nonEmptiy` property through the traversal.
the idea is to modify `Cause.Filter` to be a `Folder` producing a `(Cause[E], Boolean)` (instead of just `Cause[E]`),
I'm afraid this requires changing the signature of a public class though.

Comment (eyalfa) [reply core/shared/src/main/scala/zio/Cause.scala]: another option which greatly simplifies things **without** modifying any public signature, the filter fold can be modified to guarantee that 'filtered out' parts of the tree are replaced by `Cause.Empty`, this way the `nonEmpty` call can be replaced with `eq Cause.Empty`.
in fact the only thing preventing this from happening is the `Stackless` case, I'm not sure what's the meaning of `Stackless(Empty)` but I suspect it's equivalent to plain `Empty`.

seems to me [zio.Cause#unified](https://github.com/zio/zio/blob/e4469873546995e9351cd72d03f7e734d729c261/core/shared/src/main/scala/zio/Cause.scala#L692-L694) is the only piece of code affected by `stackless` and as far as I can read the code this function returns an empty list both for `Stackless(Empty)` and `Empty`.

Comment (eyalfa) [reply core/shared/src/main/scala/zio/Cause.scala]: > `nonEmpty` is not necessarily cheap, and it's called at least once, potentially twice per method invocation. Should we store it in vals at the start of the method for both left and right?

it's never invoked more than once on either objects, so no real gain pre calculating them

File: core/shared/src/main/scala/zio/Cause.scala (hunk 1 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/Cause.scala]: I think the fact that are already filtered is because of how `bothCase` and `thenCase` are used currently. I would be more comfortable keeping `p(c)` instead of assuming they're going to be filtered by the time we call these methods.

Comment (eyalfa) [reply core/shared/src/main/scala/zio/Cause.scala]: I actually think it's guaranteed by the fold semantics, furthermore applying `p(c)` won't be any cheaper thatn testing for emptiness.

Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/Cause.scala]: My point is that while it's guaranteed by the `fold` semantics, what's not guaranteed is that the `bothCase` and `thenCase` methods might be used in a method other than `fold`. Since `Folder.Filter` is a public class (not sure why, but we can't ignore that it is), this test will fail using the implementation in this PR:

```scala
          test("filter(false)") {
            val filter = Cause.Folder.Filter[Unit](_ => false)
            val cause = filter.bothCase((), f1, f2)
            assertTrue(cause.isEmpty)
          }
```

Comment (eyalfa) [reply core/shared/src/main/scala/zio/Cause.scala]: I'm also not sure why the `Filter` class exists as a class instead of a `fold` call or anonymous class, nevertheless seems like it's being public is simply a miss-sight.
anyway, as a `Folder` it's intended to be used by `fold`s, I guess almost any other `Folder` implementation can be abused in similar ways to your example... what matters here is the semantics of filtering a `Cause` which are currently broken and have to be fixed, I think modifying the existing code is the path of least resistance, if you disagree we can always deprecate the exiting class and create a new one or simply replace it with a private/anonymous one.

File: core/shared/src/main/scala/zio/Cause.scala (hunk 2 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/Cause.scala]: Ah yeah my bad you're right

File: core/shared/src/main/scala/zio/Cause.scala (hunk 1 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/Cause.scala]: I think for now let's go with using `p(c)` instead of `nonEmpty` in `bothCase` and `thenCase`. I'm planning on looking into the performance of error paths in the future (including `Cause`), so maybe add a `TODO` comment so that we don't forget it

**Review by kyri-petrou — COMMENTED** [review]
By the way do we need to also modify the `stacklessCase` method to be aligned with `failCase` etc?

**Comment by eyalfa** [comment]
@kyri-petrou 
I think replacing `nonEmpty` with `p(c)` is a beaking change due to multiple reasons:
1. the original code never applied `p` over a `Then` or a `Both`, doing so may introduce regressions.
2. it's quite possible these applications of `p` may fail with runtime match errors. (I suspect they are implemented using pattern matching).

**Comment by kyri-petrou** [comment]
> the original code never applied p over a Then or a Both, doing so may introduce regressions.

@eyalfa I'm not sure I understand your comment. This is the current code for `thenCase` which uses `p`. Am I misunderstanding something?

```scala
      def thenCase(context: Any, left: Cause[E], right: Cause[E]): Cause[E] =
        if (p(left)) {
          if (p(right)) Cause.Then(left, right)
          else left
        } else if (p(right)) right
        else Cause.empty
```

**Comment by eyalfaZS** [comment]
> > the original code never applied p over a Then or a Both, doing so may introduce regressions.
> 
> @eyalfa I'm not sure I understand your comment. This is the current code for `thenCase` which uses `p`. Am I misunderstanding something?
> 
> ```scala
>       def thenCase(context: Any, left: Cause[E], right: Cause[E]): Cause[E] =
>         if (p(left)) {
>           if (p(right)) Cause.Then(left, right)
>           else left
>         } else if (p(right)) right
>         else Cause.empty
> ```

it never applies `p` on the `Then` instance itself, it only considers it for composition.

**Comment by kyri-petrou** [comment]
> it never applies `p` on the `Then` instance itself, it only considers it for composition.

@eyalfa I'm not suggesting we should apply `p` on the `Then` instance itself, just `p(left)` and `p(right)` instead of `left.nonEmpty` and `right.nonEmpty` in this PR

**Comment by eyalfaZS** [comment]
> > the original code never applied p over a Then or a Both, doing so may introduce regressions.
> 
> @eyalfa I'm not sure I understand your comment. This is the current code for `thenCase` which uses `p`. Am I misunderstanding something?
> 
> ```scala
>       def thenCase(context: Any, left: Cause[E], right: Cause[E]): Cause[E] =
>         if (p(left)) {
>           if (p(right)) Cause.Then(left, right)
>           else left
>         } else if (p(right)) right
>         else Cause.empty
> ```

it would in the case of a composite `Then` (when one of the side is also a `Then`)

**Comment by jdegoes** [comment]
@eyalfa Would love to get this fix in. Looks like @kyri-petrou suggested a minor tweak, and then looks good to merge!

**Comment by eyalfaZS** [comment]
@jdegoes , @kyri-petrou ,
sorry for being less responsive the past few days, I've taken a new position and last few days were packed.

I think Kyrie's suggestion changes the semantics of the `filter` in such a way that it may break existing code.
original implementation applied `p` only on the 'leafs', never on the `Both` and `Then` cases (`Stackless` neither), Kyrie's suggestion does that, i.e. when we have something like `Both(c0, Both(c1, c2))` `p` would be applied on the nested `Both` as well which is likely to fail if `p` is implemented as a partial function.

looking at the original code it's quite clear that `Both`, `Then` and `Stackless` are treated as containers/collections of causes and are never filtered on their own. I think a better approach is to replaced filtered out causes with a `Cause.Empty` and pattern match when composing - which is very similar to what the current code is doing (with a bug 😎 ).

* another issue I see with this operator is losing annotations, seems like there are two 'levels' of implementing a `Fold` and for some reason `Filter` chose the one losing annotations.

**Comment by eyalfa** [comment]
@kyri-petrou I think this is the best way to go, if you're still concerned about anyone using this class directly we can always deprecate it and replace it with a new one or just use the fold override taking lambdas

**Comment by kyri-petrou** [comment]
> I think Kyrie's suggestion changes the semantics of the filter in such a way that it may break existing code.
original implementation applied p only on the 'leafs', never on the Both and Then cases (Stackless neither)

@eyalfa This is not true. In fact, what's changing the semantics of filter is the new implementation. I managed to create a reproducer that showcases a change in behaviour with your implementation:

```scala
object App {
  val c1   = Cause.fail("foo")
  val c2   = Cause.fail("bar")
  val c3   = Cause.fail("baz")
  val c23  = Cause.Both(c2, c3)
  val c123 = Cause.Both(c1, c23)

  def main(args: Array[String]): Unit = {
    val c = c123.filter { c =>
      println(s"applying filter on: ${c.failures}")
      !c.isInstanceOf[Cause.Both[?]]
    }

    println(s"\nfiltered cause: $c")
  }
}
```

This code yields the following outputs:

**series/2.x:**
```
applying filter on: List(bar)
applying filter on: List(baz)
applying filter on: List(foo)
applying filter on: List(bar, baz)

filtered cause: Fail(foo,Stack trace for thread "zio-fiber-":
)
```

**PR (current `bothCase` implementation)**
```
applying filter on: List(foo)
applying filter on: List(bar)
applying filter on: List(baz)

filtered cause: Both(Fail(foo,Stack trace for thread "zio-fiber-":
),Both(Fail(bar,Stack trace for thread "zio-fiber-":
),Fail(baz,Stack trace for thread "zio-fiber-":
)))
```

**PR (/w my suggestion)**
```
applying filter on: List(foo)
applying filter on: List(bar)
applying filter on: List(baz)
applying filter on: List(bar)
applying filter on: List(baz)
applying filter on: List(foo)
applying filter on: List(bar, baz)

filtered cause: Fail(foo,Stack trace for thread "zio-fiber-":
)
```

As you can see, both the current code (series/2.x) and my suggestion apply the filter on `Both`, but that's not the case with your proposed changes.

Having said that, the downside of my suggestion is that it ends up evaluating p(c) on same Cause multiple times, which is suboptimal. There are ways to improve performance in this case, but since I find it extremely difficult to picture `Cause#filter` being used in CPU hotpaths, I think we can ignore the performance penalty for now. In the future I plan to visit `Cause` methods and optimize them as much as possible

**Comment by eyalfa** [comment]
@kyri-petrou ,
I actually understood that the original impl also applies p to `Both` and `Then` in the nested case, unfortunately I wrote my comment before this understanding hitting me 😎 .

I think at this point it's a matter of deciding what's the correct behavior, I suspect it's applying p on the `Both` case only when one is actually constructed, but tbh I'm not 100% sure.

another thing is, I wouldn't expect filter to apply the predicate more than once per `Cause`, original impl has this property. but it includes a bug for the 'single' case.

I think the logic for `Both` should be something like:
if both sides are filtered out return Empty
if exactly one side is filtered out, return the other
otherwise, construct a `Both`, apply the predicate on the new instance, in case of false return Empty otherwise return the new instance.

sketched the implementation and got these printouts when running your sample:
```
applying filter on: List(foo)
applying filter on: List(bar)
applying filter on: List(baz)
applying filter on: List(bar, baz)

filtered cause: Fail(foo,Stack trace for thread "zio-fiber-":
)
```

I think this approach yields the correct results while maintaining the 'apply once' property.

**Comment by eyalfaZS** [comment]
@kyri-petrou ☝️

File: core/shared/src/main/scala/zio/Cause.scala [hunk 6]
```diff
@@ -822,28 +822,64 @@ object Cause extends Serializable {
...
+          Cause.empty
+      }
 
       def bothCase(context: Any, left: Cause[E], right: Cause[E]): Cause[E] =
-        if (p(left)) {
-          if (p(right)) Cause.Both(left, right)
-          else left
-        } else if (p(right)) right
-        else Cause.empty
+        (left, right) match {
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/Cause.scala]: Let's avoid the intermediate tuple creation here and in the methods below and use simple `if (left eq Cause.Empty) { ... } else if (right eq Cause.Empty) { ... } else {  ...  }` statements

Comment (eyalfaZS) [reply core/shared/src/main/scala/zio/Cause.scala]: done

**Review by kyri-petrou — APPROVED** [review]
